File: soloq/submissions/accepted/soloq.py
from collections import deque

MOD = 1000000007

class AhoCorasick:

    class Node:
        def __init__(self):
            self.to = [-1,-1]
            self.link = -1
            self.ends_here = 0
            self.end = False
            self.terminal = 0

    def __init__(self):
        self.d = [self.Node()]

    
    def insert(self, s, i):
        v = 0
        for C in s:
            c = 1 if C == 'L' else 0
            if self.d[v].to[c] == -1:
                self.d[v].to[c] = len(self.d)
                self.d.append(self.Node())
            v = self.d[v].to[c]
        self.d[v].end = True
        self.d[v].ends_here += 1
    
    def push_links(self):
        self.d[0].link = -1
        q = deque()
        q.append(0)
        while q:
            v = q.popleft()
            if self.d[v].link == -1:
                self.d[v].terminal = 0
            elif self.d[self.d[v].link].end:
                self.d[v].terminal = self.d[v].link
            else:
                self.d[v].terminal = self.d[self.d[v].link].terminal
            self.d[v].ends_here += self.d[self.d[v].terminal].ends_here
            
            for i in range(2):
                u = self.d[v].to[i]
                if u == -1:
                    continue
                self.d[u].link = 0 if self.d[v].link == -1 else self.d[self.d[v].link].to[i]
                q.append(u)
            
            if v != 0:
                for i in range(2):
                    if self.d[v].to[i] == -1:
                        self.d[v].to[i] = self.d[self.d[v].link].to[i]




def matMul(A, B):
    N, L, M = len(A), len(A[0]), len(B[0])
    res = [[0 for _ in range(M)] for _ in range(N)]
    for i in range(N):
        for j in range(M):
            for k in range(L):
                res[i][j] += A[i][k] * B[k][j]
                res[i][j] %= MOD
    return res

def matSum(A, B):
    return [[(A[i][j] + B[i][j]) % MOD for j in range(len(A[0]))] for i in range(len(A))]

def eye(n):
    return [[0 if j != i else 1 for j in range(n)] for i in range(n)]

def matPow(A, b):
    res = eye(len(A))
    while b != 0:
        if b % 2 == 1:
            res = matMul(res, A)
        A = matMul(A, A)
        b //= 2
    return res
    

def solve(N: int, K: int, X: int, Y: int, W1: int, W2: int, L1: int, L2: int, S: list[str]) -> int:

    # Transform the chances to Fp
    W1 *= pow(W2, MOD - 2, MOD)
    L1 *= pow(L2, MOD - 2, MOD)


    # Construct graph
    AC = AhoCorasick()

    for i in range(len(S)):
        AC.insert(S[i], i)
    
    AC.push_links()

    num_states = len(AC.d)

    # Construct transition matrix
    A = [[0 for _ in range(2 * num_states)] for _ in range(2 * num_states)]
    
    for i in range(num_states):
        j_win = AC.d[i].to[0]
        swap_win = AC.d[j_win].ends_here % 2 == 1
        if swap_win:
            A[i][j_win + num_states] = W1
            A[i + num_states][j_win] = L1
        else:
            A[i][j_win] = W1
            A[i + num_states][j_win + num_states] = L1
        
        j_lose = AC.d[i].to[1]
        swap_lose = AC.d[j_lose].ends_here % 2 == 1
        if swap_lose:
            A[i][j_lose + num_states] = MOD - W1
            A[i + num_states][j_lose] = MOD - L1
        else:
            A[i][j_lose] = MOD - W1
            A[i + num_states][j_lose + num_states] = MOD - L2
    
    d0 = [[0 if i != 0 else 1 for i in range(2 * num_states)]]
    mu = [[X * W1 - Y * (MOD - W1) if i < num_states else X * L1 - Y * (MOD - L1)] for i in range(2 * num_states)]
    for v in mu:
        v[0] = (v[0] % MOD + MOD) % MOD
    
    # Returns sum from i = 1 to n of A^i
    def fun(n):
        if n == 1:
            return A
        elif n % 2 == 1:
            return matSum(matPow(A, n), fun(n - 1))
        else:
            return matMul(matSum(eye(len(A)), matPow(A, n // 2)), fun(n // 2))
    
    return matMul(d0, matMul(matSum(eye(len(A)), fun(N - 1)), mu))[0][0]



def main():
    T = int(input())
    for _ in range(T):
        N, K = map(int, input().split())
        X, Y, W1, W2, L1, L2 = map(int, input().split())
        S = [input() for _ in range(K)]
        print(solve(N, K, X, Y, W1, W2, L1, L2, S))

if __name__ == '__main__':
    main()

File: soloq/submissions/accepted/soloq.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 1000000007;

using ll = long long;

template<int MOD, int RT> struct mint {
	static const int mod = MOD;
	static constexpr mint rt() { return RT; } // primitive root for FFT
	int v; explicit operator int() const { return v; } // explicit -> don't silently convert to int
	mint():v(0) {}
	mint(ll _v) { v = int((-MOD < _v && _v < MOD) ? _v : _v % MOD);
		if (v < 0) v += MOD; }
	bool operator==(const mint& o) const {
		return v == o.v; }
	friend bool operator!=(const mint& a, const mint& b) { 
		return !(a == b); }
	friend bool operator<(const mint& a, const mint& b) { 
		return a.v < b.v; }

	mint& operator+=(const mint& o) { 
		if ((v += o.v) >= MOD) v -= MOD; 
		return *this; }
	mint& operator-=(const mint& o) { 
		if ((v -= o.v) < 0) v += MOD; 
		return *this; }
	mint& operator*=(const mint& o) { 
		v = int((ll)v*o.v%MOD); return *this; }
	mint& operator/=(const mint& o) { return (*this) *= inv(o); }
	friend mint pow(mint a, ll p) {
		mint ans = 1; assert(p >= 0);
		for (; p; p /= 2, a *= a) if (p&1) ans *= a;
		return ans; }
	friend mint inv(const mint& a) { assert(a.v != 0); 
		return pow(a,MOD-2); }
		
	mint operator-() const { return mint(-v); }
	mint& operator++() { return *this += 1; }
	mint& operator--() { return *this -= 1; }
	friend mint operator+(mint a, const mint& b) { return a += b; }
	friend mint operator-(mint a, const mint& b) { return a -= b; }
	friend mint operator*(mint a, const mint& b) { return a *= b; }
	friend mint operator/(mint a, const mint& b) { return a /= b; }
};

using mi = mint<MOD, 5>;

using Mat = vector<vector<mi>>;

Mat makeMat(int r, int c) { return Mat(r, vector<mi>(c)); }
Mat makeId(int n) {
    Mat m = makeMat(n, n);
    for (int i = 0; i < n; ++i)
        m[i][i] = 1;
    return m;
}
Mat& operator+=(Mat& a, Mat const& b) {
    assert(size(a) == size(b) && size(a[0]) == size(b[0]));
    for (int i = 0; i < size(a); ++i)
        for (int j = 0; j < size(a[0]); ++j)
            a[i][j] += b[i][j];
    return a;
}
Mat operator+(Mat a, Mat const& b) { return a += b; }

Mat operator*(Mat const& a, Mat const& b) {
    int x = size(a), y = size(a[0]), z = size(b[0]);
    assert(y == size(b));
    Mat c = makeMat(x, z);
    for (int i = 0; i < x; ++i)
        for (int j = 0; j < y; ++j)
            for (int k = 0; k < z; ++k)
                c[i][k] += a[i][j] * b[j][k];
    return c;
}
Mat& operator*=(Mat& a, Mat const& b) { return a = a * b; }
Mat pow(Mat m, long long p) {
    int n = size(m);
    assert(n == size(m[0]) && p >= 0);
    Mat res = makeId(n);
    for (; p; p /= 2, m *= m) if (p & 1) res *= m;
    return res;
}
Mat geometricSum(Mat const& m, long long n) {
    if (n == 0) return makeId(size(m));
    // case 2k
    if (n % 2 == 0) return pow(m, n) + geometricSum(m, n - 1);
    // case 2k - 1 with n / 2 = k - 1 
    return (makeId(size(m)) + pow(m, n / 2)) * geometricSum(m, n / 2);
}

struct AhoCorasick {
    struct Node {
        int link;
        vector<int> to = {0,0};
        int ends_here = 0;
        bool end;
        int terminal;
    };
    vector<Node> d{{}};
    void insert(string const& s, int i) {
        int v = 0;
        for (char C : s) {
            int c = C == 'L';
            if (!d[v].to[c]) {
                d[v].to[c] = int(size(d));
                d.emplace_back();
            }
            v = d[v].to[c];
        }
        d[v].end = true;
        d[v].ends_here = 1;
    }

    void push_links() {
        d[0].link = -1;
        queue<int> q;
        q.push(0);
        while (!q.empty()) {
            int v = q.front(); q.pop();
            d[v].terminal = d[v].link == -1 ? 0 : d[d[v].link].end ? d[v].link : d[d[v].link].terminal;
            d[v].ends_here += d[d[v].terminal].ends_here;
            for (int c = 0; c < 2; ++c) {
                int u = d[v].to[c];
                if (!u) continue;
                d[u].link = d[v].link == -1 ? 0 : d[d[v].link].to[c];
                q.push(u);
            }
            if (v) {
                for (int c = 0; c < 2; ++c) {
                    if (!d[v].to[c]) {
                        d[v].to[c] = d[d[v].link].to[c];
                    }
                }
            }
        }
    }

};


/**
 * Return the sum of A and B.
 * 
 * A: a non-negative integer
 * B: another non-negative integer
 */
int solve(int N, int K, int X, int Y, int W1, int W2, int L1, int L2, vector<string>& S) {
    // Transform chances to Fp
    mi W = mi(W1) / mi(W2);
    mi L = mi(L1) / mi(L2);

    // Build graph
    AhoCorasick AC;
    for (int i = 0; i < K; ++i) {
        AC.insert(S[i],i);
    }
    AC.push_links();
    int n = size(AC.d);

    // Build transition matrix
    vector<vector<mi>> A(2 * n, vector<mi>(2 * n, 0));
    for (int i = 0; i < n; ++i) {
        int j_win = AC.d[i].to[0];
        bool swap_win = AC.d[j_win].ends_here & 1;
        if (swap_win) {
            A[i][j_win + n] = W;
            A[i + n][j_win] = L;
        } else {
            A[i][j_win] = W;
            A[i + n][j_win + n] = L;
        }
        int j_lose = AC.d[i].to[1];
        bool swap_lose = AC.d[j_lose].ends_here & 1;
        if (swap_lose) {
            A[i][j_lose + n] = mi(0) - W;
            A[i + n][j_lose] = mi(0) - L;
        } else {
            A[i][j_lose] = mi(0) - W;
            A[i + n][j_lose + n] = mi(0) - L;
        }
    }

    vector<vector<mi>> d0(1, vector<mi>(2 * n, 0));
    d0[0][0] = 1;
    vector<vector<mi>> mu(2 * n, vector<mi>(1));
    for (int i = 0; i < n; ++i) {
        mu[i][0] = mi(X) * W - mi(Y) * (mi(0) - W);
        mu[i+n][0] = mi(X) * L - mi(Y) * (mi(0) - L);
    }

    auto ans = d0 * geometricSum(A, n - 1) * mu;
    return ans[0][0].v;

}

int main() {
    int T;
    cin >> T;
    for (int i = 0; i < T; i++) {
        int N, K, X, Y, W1, W2, L1, L2;
        cin >> N >> K;
        cin >> X >> Y >> W1 >> W2 >> L1 >> L2;
        vector<string> S(K);
        for (int j = 0; j < K; ++j) {
            cin >> S[j];
        }
        cout << solve(N, K, X, Y, W1, W2, L1, L2, S) << '\n';
    }
}
